### PR TITLE
chore: Update Raft migrate tool to handle new proposal format

### PR DESCRIPTION
**Pull request by rahulgurnani**
Fixes DGRAPH-2832
This PR updates the raft wal migrate tool. We recently updated raft proposal format for alpha and zero, see #7059 .
The tool now can migrate raftwal from old proposal format to new proposal format. 
**This change will help to migrate raftwal from RC-2 to RC-4 onwards.**

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)

### CODE REVIEW

File: dgraph/cmd/raft-migrate/run.go [hunk 1]
```diff
@@ -68,29 +131,15 @@ func run(conf *viper.Viper) error {
...
-		log.Fatalf("Failed to read encryption file: %s", err)
-	}
-
-	// Copied over from zero/run.go
-	kvOpt := badger.LSMOnlyOptions(oldDir).
-		WithSyncWrites(false).
-		WithValueLogFileSize(64 << 20).
-		WithIndexCacheSize(100 << 20).
-		WithEncryptionKey(encKey)
+	// TODO(rahul): Maybe uncomment following
```
Comment (rahulgurnani) [inline dgraph/cmd/raft-migrate/run.go:134]: I think this needs to handled

**Comment by anurags92** [comment]
> This is to unable migrating from 20.11-rc3 to 20.11.-rc4 without backup restore.
> 
> This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)

This is to enable migration or disable migration without backup restore? The comment is unclear without context.

**Review by rahulgurnani — COMMENTED** [review]
Updated the comment, hope it makes sense now.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-:-MOHhKrSBUZrKJFO6QK_:b-btt7gw)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @jarifibrahim, @manishrjain, @NamanJain8, and @vvbalaji-dgraph)

**Review by manishrjain — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MORrQDiEmeUbXbg2z6O:b27y6yz) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
> ```Go
> )
> 
> type Proposal struct {
> ```

Don't need this file. Protobufs are backwards compatible.
___
*[dgraph/cmd/raft-migrate/run.go, line 82 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrVL71GUlM-Q_VN-A:-MORrVL71GUlM-Q_VN-B:b-8xcfhq) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/run.go#L82)):*
> ```Go
> 	newKey := parseAndConvertKey("%02d-%d", oldProposal.Key)
> 	var newProposal pb.Proposal
> 	newProposal.Mutations = oldProposal.Mutations
> ```

Don't need this either. Protobufs are compatible.

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 3 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @NamanJain8, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MOWst6SCB6rAoj8CsWn:b-q68dnu) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need this file. Protobufs are backwards compatible.
</blockquote></details>

We don't have `Key` in the new Proposal. Hence, we need the old types(and their unmarshaller) to unmarshal into. 
___
*[dgraph/cmd/raft-migrate/run.go, line 82 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrVL71GUlM-Q_VN-A:-MOWtIh38p5OWfVHGK-a:b821llw) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/run.go#L82)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need this either. Protobufs are compatible.
</blockquote></details>

We need all this because of  `Key`.

**Review by manishrjain — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/debug/wal.go, line 51 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWwkPr-rbUFq2QmrGJ:-MOWwkPr-rbUFq2QmrGK:b-3pfph3) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/debug/wal.go#L51)):*
> ```Go
> 	var err error
> 	if isZero {
> 		var zpr pb.ZeroProposal
> ```

The previous code was fine. No need to change this, or add isZero.
___
*[dgraph/cmd/debug/wal.go, line 125 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWwr926BqTPko5yfLE:-MOWwr926BqTPko5yfLF:b-q0532v) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/debug/wal.go#L125)):*
> ```Go
> }
> 
> func printRaft(db *badger.DB, store *raftmigrate.OldDiskStorage, groupId uint32) {
> ```

No need to change.
___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MOWwxwT3HecHB78bsOQ:b-rdw9b7) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

We don't have `Key` in the new Proposal. Hence, we need the old types(and their unmarshaller) to unmarshal into. 
</blockquote></details>

But, that id is reserved still. It should work. Did you try?

**Review by ahsanbarkati — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/run.go, line 162 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWymvh3ruBfOUqYWB2:-MOWymviDzpnxbHOt_x9:b-wd3xw3) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/raft-migrate/run.go#L162)):*
> ```Go
> 	fmt.Printf("Fetching entries from low: %d to high: %d\n", firstIndex, lastIndex)
> 	// Should we batch this up?
> 	oldEntries, err := oldWal.Entries(1, lastIndex+1, math.MaxUint64)
> ```

should be `oldWal.Entries(firstIndex, lastIndex+1, math.MaxUint64)`
___
*[dgraph/cmd/raft-migrate/run.go, line 196 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWz1WG6Omypek8Go5Q:-MOWz1WH6csiTFf3gM5X:by0iq0a) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/raft-migrate/run.go#L196)):*
> ```Go
> 	newWal, err := raftwal.InitEncrypted(newDir, encKey)
> 	x.Check(err)
> ```

`defer newWal.Close()`

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/debug/wal.go, line 51 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWwkPr-rbUFq2QmrGJ:-MOWyWiuCeqO3V9mFXMZ:ba4dytm) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/debug/wal.go#L51)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

The previous code was fine. No need to change this, or add isZero.
</blockquote></details>

We noticed that the previous code was not printing out the correct entries for ZeroProposal was able to get unmarshalled into Proposal. Hence, always the first statement was getting executed. 
___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MOX0sIC9qvDIYHyfwGC:bn90riw) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

But, that id is reserved still. It should work. Did you try?
</blockquote></details>

Yes, I tried retrieving `Key` from the data. But I couldn't find a way unless we write our own unmarhsaller.
___
*[dgraph/cmd/raft-migrate/run.go, line 162 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWymvh3ruBfOUqYWB2:-MOX1bAA4Drd3-7gre88:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/raft-migrate/run.go#L162)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

should be `oldWal.Entries(firstIndex, lastIndex+1, math.MaxUint64)`
</blockquote></details>

Done.
___
*[dgraph/cmd/raft-migrate/run.go, line 196 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MOWz1WG6Omypek8Go5Q:-MOX1_Qs4k6pSN6GjzTB:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/e7221ad62b35c591166a1b528424702b40c01689/dgraph/cmd/raft-migrate/run.go#L196)):*
<details><summary><i>Previously, ahsanbarkati (Ahsan Barkati) wrote…</i></summary><blockquote>

`defer newWal.Close()`
</blockquote></details>

Done.

**Review by manishrjain — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MO_Bp977xPMMnniOg7n:by9vh6q) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

Yes, I tried retrieving `Key` from the data. But I couldn't find a way unless we write our own unmarhsaller.
</blockquote></details>

You don't need to retrieve the key from the data. It's not *that* important. Does something not work without it?

**Review by NamanJain8 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @manishrjain, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MO_GQrF6aqdCyJqWap_:b-k5hgt) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need to retrieve the key from the data. It's not *that* important. Does something not work without it?
</blockquote></details>

If we don't have a key, then every proposal in raftwal would be considered as invalid. That would mean that we are starting from a fresh state and ignoring the current raftwal https://github.com/dgraph-io/dgraph/blob/11e46217bc09e0e961e167c0a86e427f6f1ff046/dgraph/cmd/zero/raft.go#L336-L338

File: dgraph/cmd/raft-migrate/run.go [hunk 2]
```diff
@@ -50,13 +50,22 @@ func init() {
...
 }
 
+func updateEntry(entry raftpb.Entry) raftpb.Entry {
+	// Raft commits an empty entry on becoming leader.
+	if entry.Type == raftpb.EntryConfChange || len(entry.Data) == 0 {
+		return entry
+	}
+
+	data := make([]byte, 8+len(entry.Data))
+	copy(data[8:], entry.Data)
```
Comment (jarifibrahim) [inline dgraph/cmd/raft-migrate/run.go:64]: Don't you need to set the key on the first 8 bytes?

Comment (jarifibrahim) [reply dgraph/cmd/raft-migrate/run.go:64]: Add a comment if we don't need the key

**Review by manishrjain — APPROVED** [review]
Do we still have the Badger code? We should remove it.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-:-MOgpBKy3khrQHvBT8Qb:b-vjas7g)* status: 0 of 4 files reviewed, 8 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

___
*[dgraph/cmd/raft-migrate/generated.go, line 31 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-MORrQDiEmeUbXbg2z6N:-MOgp1Zk3jopbeb0OKgN:bz2i9w0) ([raw file](https://github.com/dgraph-io/dgraph/blob/5b46dec97182acc01579fb1ab27bbbce3b5fba00/dgraph/cmd/raft-migrate/generated.go#L31)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

If we don't have a key, then every proposal in raftwal would be considered as invalid. That would mean that we are starting from a fresh state and ignoring the current raftwal https://github.com/dgraph-io/dgraph/blob/11e46217bc09e0e961e167c0a86e427f6f1ff046/dgraph/cmd/zero/raft.go#L336-L338
</blockquote></details>

You can remove that code from Zero.

**Review by NamanJain8 — COMMENTED** [review]
Yes, the code in raft-migrate/storage.go should be removed. This code is currently used by dgraph debug tool for debugging old badger wal. @ajeetdsouza is working on it. This code removal can be part of separate PR.  

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7123#-:-MOgpj2_AaXqMsC8JZbc:bywibm5)* status: 0 of 4 files reviewed, 8 unresolved discussions (waiting on @ahsanbarkati, @manishrjain, @NamanJain8, @rahulgurnani, and @vvbalaji-dgraph)

File: dgraph/cmd/raft-migrate/run.go (hunk 2 above)
Comment (NamanJain8) [reply dgraph/cmd/raft-migrate/run.go:68]: Done

**Pull request merged**